### PR TITLE
Filter support for reflog and streamUtil implementation

### PR DESCRIPTION
**Pull request by ajantha-bhat**
For new GC algorithm, need to filter just DELETE_REFERENCE and ASSIGN_REFERENCE reflog entries from the backend store. Having CEL filter can help in reducing the data in HTTP response by providing the filtered results.

### CODE REVIEW

**Comment by ajantha-bhat** [comment]
> For new GC algorithm, need to filter just DELETE_REFERENCE and ASSIGN_REFERENCE reflog entries from the backend store. Having CEL filter can help in reducing the data in HTTP response by providing the filtered results.

@snazy : I know I got a comment to remove filtering during reflog implementation. But now the GC algorithm demands it and it can be one of the usecase for filtering. Hence need this PR.

File: model/src/main/java/org/projectnessie/api/params/RefLogParams.java [hunk 1]
```diff
@@ -52,16 +52,30 @@
...
+              + "Hint: when filtering entries, you can determine whether entries are \"missing\" (filtered) by "
+              + "checking whether 'ReflogResponseEntry.parentRefLogId' is different from the hash of the previous "
+              + "reflog in the log response.")
+  @QueryParam("filter")
+  private String filter;
+
   public RefLogParams() {}
 
-  private RefLogParams(String startHash, String endHash, Integer maxRecords, String pageToken) {
+  private RefLogParams(
```
Comment (snazy) [inline model/src/main/java/org/projectnessie/api/params/RefLogParams.java:69]: The whole class is missing a `RefLogParamsTest`, similar to the other params test classes.

File: model/src/main/java/org/projectnessie/api/params/RefLogParams.java [hunk 2]
```diff
@@ -52,16 +52,30 @@
...
+      description =
+          "A Common Expression Language (CEL) expression. An intro to CEL can be found at https://github.com/google/cel-spec/blob/master/doc/intro.md.\n\n"
+              + "Usable variables within the expression are:\n\n"
+              + "- 'reflog' with fields 'refLogId' (string), 'refName' (string), 'commitHash' (string), 'parentRefLogId' "
+              + "(string), ',operation' (string), 'operationTime' (long)\n\n"
+              + "Hint: when filtering entries, you can determine whether entries are \"missing\" (filtered) by "
+              + "checking whether 'ReflogResponseEntry.parentRefLogId' is different from the hash of the previous "
+              + "reflog in the log response.")
+  @QueryParam("filter")
+  private String filter;
```
Comment (snazy) [inline model/src/main/java/org/projectnessie/api/params/RefLogParams.java:65]: `@Nullable` missing

File: clients/client/src/main/java/org/projectnessie/client/StreamingUtil.java [hunk 3]
```diff
@@ -114,6 +115,34 @@ private StreamingUtil() {
...
+   * Default implementation to return a stream of reflog entries, functionally equivalent to calling
+   * {@link NessieApiV1#getRefLog()} with manual paging.
+   *
+   * <p>The {@link Stream} returned by {@code getReflogStream(OptionalInt.empty())}, if not limited,
+   * returns all reflog entries.
+   *
+   * @param api The {@link NessieApiV1} to use
+   * @return stream of {@link RefLogResponse.RefLogResponseEntry} objects
+   */
+  public static Stream<RefLogResponse.RefLogResponseEntry> getReflogStream(
```
Comment (snazy) [inline clients/client/src/main/java/org/projectnessie/client/StreamingUtil.java:128]: No tests for this?

**Review by snazy — COMMENTED** [review]
Can you elaborate _why_ the filter is needed now?

File: clients/client/src/main/java/org/projectnessie/client/StreamingUtil.java (hunk 3 above)
Comment (ajantha-bhat) [reply clients/client/src/main/java/org/projectnessie/client/StreamingUtil.java:128]: the filter testcase is using this streamUtil itself. So, it is covered.

**Comment by ajantha-bhat** [comment]
> Can you elaborate why the filter is needed now?
As mentioned in the GC algorithm document (shared internally long back). Originally I thought I will get all the commits from reflog and do anti join with LIVE commits to get expired commits. But finding live commits is no use as the commit can have both dead and live contents. So, the granularity should be contents not commits. So, dropped this algorithm. 

Instead changed algorithm to fetch dead contents using reflog by traversing only DELETE_REF and ASSIGN_REF (potentially unreachable) and union them with expired contents form live reference. Hence need filter to fetch reflog entries for only DELETE_REF and ASSIGN_REF.

File: model/src/main/java/org/projectnessie/api/params/RefLogParams.java (hunk 2 above)
Comment (ajantha-bhat) [reply model/src/main/java/org/projectnessie/api/params/RefLogParams.java:65]: it is missed in commitLogParams also, I will handle it.

File: model/src/main/java/org/projectnessie/api/params/RefLogParams.java (hunk 1 above)
Comment (ajantha-bhat) [reply model/src/main/java/org/projectnessie/api/params/RefLogParams.java:70]: Added it similar to `CommitLogParamsTest`

**Review by nastra — APPROVED** [review]

**Pull request merged**